### ISSUE TITLE
Segfault with tcp-check connect with threads (haproxy 1.9.x)

### ISSUE THREAD

**Issue by renard** [haproxy/haproxy#148]
## Output of `haproxy -vv` and `uname -a`

`haproxy -vv`
```
HA-Proxy version 1.9.8-1~bpo9+1 2019/05/18 - https://haproxy.org/
Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-1.9.8=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_NS=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
Running on OpenSSL version : OpenSSL 1.1.0f  25 May 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.22 2016-07-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with multi-threading support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE
              h2 : mode=HTTP       side=FE
       <default> : mode=HTX        side=FE|BE
       <default> : mode=TCP|HTTP   side=FE|BE

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

```

`uname -a`

```
Linux test-node 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u2 (2017-06-26) x86_64 GNU/Linux
```

## What's the configuration?

```
global
	log /dev/log	local0
	log /dev/log	local1 notice
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
	stats timeout 30s
	user haproxy
	group haproxy
	daemon
        # Threads are only here to trigger the segfault. If not used, no segfault so far.
        nbthread 32

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	# Default ciphers to use on SSL-enabled listening sockets.
	# For more information, see ciphers(1SSL). This list is from:
	#  https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/
	# An alternative list with additional directives can be obtained from
	#  https://mozilla.github.io/server-side-tls/ssl-config-generator/?server=haproxy
	ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
	ssl-default-bind-options no-sslv3

defaults
	log	global
	mode	http
	option	httplog
	option	dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http
```

You can generate other backends using following script (put it in `/etc/haproxy/genconfig`):

```
#!/bin/bash

cd /etc/haproxy
mkdir -p sites-enabled

rm sites-enabled/*

cat <<EOF > sites-enabled/00-stats.cfg
listen stats
    bind *:80
    stats uri /
    stats admin if TRUE
EOF

for i in $(seq 20); do
cat <<EOF > sites-enabled/front-$i.cfg
listen frontend$i
        mode tcp
        option tcplog

        bind 127.0.0.$i:12345
        #bind 127.0.1.$i:12345

        option tcp-check
        tcp-check connect port 25
        tcp-check expect rstring "^220 "
        tcp-check send "quit\r\n"

        server 127.0.0.1 127.0.0.1:25 check inter 1000ms
        #server 127.0.1.1 127.0.1.1:25 check inter 100ms
EOF

done
```

Just make sure to add following in  `/etc/default/haproxy` on a Debian system

```
EXTRAOPTS="-f /etc/haproxy/sites-enabled/"
```

To start haproxy:

```
/etc/init.d/haproxy start
```

## Steps to reproduce the behavior

1. Install a mail server (because the example uses port 25 but can be reproduced on any port). Exim or postfix would do the job.
2. Generate backends using the provided script above.
3. See segfault using command: `dmesg -T -w | grep haproxy`

Segfault lines looks like:
```
[Tue Jul  2 17:33:19 2019] haproxy[25613]: segfault at 6 ip 00005585fd55e203 sp 00007eff467d92f0 error 4 in haproxy[5585fd48e000+1e6000]
[Tue Jul  2 17:34:56 2019] haproxy[25720]: segfault at 10 ip 0000559044aa9e85 sp 00007fb7c3fdc390 error 4 in haproxy[5590449d8000+1e6000]
```

## Actual behavior

When `tcp-check connect` is activated, haproxy segfault. Google showed a lot of people having the same issue and saying the bug has been fixed in version 1.9.6 or later versions (https://www.mail-archive.com/haproxy@formilux.org/msg33141.html).

However I check with the last available version at https://haproxy.debian.net/ and still have the issue.

I have made several tests and it turns out that the more frontends / backends you have the more frequently the segfault occurs. With 4 frontends I had this about 4 times a day, with over 100 frontend it happens every 2 or 3 minutes event if only one of them is using the `tcp-check connect`.

If the custom tcp check is commented out, (only L4 checks) everything works like a charm.

## Expected behavior

No segfault even when using tcp checks.

## Do you have any idea what may have caused this?

No. Maybe related to this bug https://www.mail-archive.com/haproxy@formilux.org/msg33141.html?

## Do you have an idea how to solve the issue?

A workaround is to not use tcp checks but this is not a solution.

I can provide further information if required.

Thanks

**Comment by cognet** [comment]
Hi,

As it so happens, I just fixed something that is very similar to what's you're describing.
Can you try the attached patch and let me know if it fixes it for you ?

Thanks !
[0001-WIP-tcp-checks-in-1.9.txt](https://github.com/haproxy/haproxy/files/3350916/0001-WIP-tcp-checks-in-1.9.txt)

**Comment by renard** [comment]
Hi,

Thanks for the quick response.

I will build a new version. Since I use @vincentbernat version I didn't build anything.
I come back to you tomorrow.

Thanks.

**Comment by renard** [comment]
Hi @cognet,

I rebuild @vincentbernat Haproxy package and added your patch into his patches list.

So far so good no segfault for the last 10 minutes. I can suppose this fixes the issue. However I'll wait for a few more hours to make sure the issue is really fixed.



Thanks again.

**Comment by renard** [comment]
After 3 hours no segfault.

I bet this solves the issue. Now just need to wait until this patch is applied into upstream and debian packages.


Thanks

**Comment by cognet** [comment]
Thanks a lot ! It should be part of the 1.9.9 release, I'm sure @vincentbernat will update his package :)

**Comment by capflam** [comment]
Just for the record, the issue was fixed by the commit f49495798
